### PR TITLE
feat(gateway): IPFSBackend  metrics and HTTP range support

### PR DESCRIPTION
**Pull request by lidel**
This PR is part of https://github.com/ipfs/bifrost-gateway/pull/61

- Moves metric-related code to `gateway/metrics.go`
- Adds opentelemetry span for every IPFSBackend API call 
- Adds success/failure histogram for each API call ipfs_gw_backend_api_call_duration_seconds 

## TODO

- [x] `Get` and `GetRange` are now the same, as suggested in https://github.com/ipfs/boxo/pull/240#pullrequestreview-1365554977
- [x] fix tests: check why Kubo sharness fails
  - [x] Kubo sharness / CI  in https://github.com/ipfs/kubo/pull/9786 must be green

### CODE REVIEW

File: gateway/metrics.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+	begin := time.Now()
+	name := "IPFSBackend.GetCAR"
+	ctx, span := spanTrace(ctx, name, trace.WithAttributes(attribute.String("path", path.String())))
+	defer span.End()
+
+	md, rc, errCh, err := b.api.GetCAR(ctx, path)
+
+	// TODO: handle errCh
+	b.updateApiCallMetric(name, err, begin)
+	return md, rc, errCh, err
```
Comment (lidel) [inline gateway/metrics.go:140]: - [ ] TODO

Comment (lidel) [inline gateway/metrics.go:128]: :thought_balloon: (late friday wrap-up)  @aschmahmann  `updateApiCallMetric` will increase counter for success or failure based on `err`. Are we missing anything critical if we only check `err` here and ignore `errCh`? 
(This should not block this PR, we can fix metric later)

File: blockservice/blockservice.go [hunk 2]
```diff
@@ -244,7 +244,7 @@ func getBlock(ctx context.Context, c cid.Cid, bs blockstore.Blockstore, fget fun
 
 		// TODO be careful checking ErrNotFound. If the underlying
 		// implementation changes, this will break.
-		logger.Debug("Blockservice: Searching bitswap")
+		logger.Debug("BlockService: Searching")
```
Comment (lidel) [inline blockservice/blockservice.go:247]: :information_source: there is no bitswap in bifrost-gateway, yet we got super confusing "Searching bitswap" message here ;)

File: gateway/metrics.go
Comment (lidel) [inline gateway/metrics.go:1]: :information_source:  moved metrics code here

File: gateway/gateway.go [hunk 3]
```diff
@@ -86,15 +88,24 @@ func NewGetResponseFromDirectoryListing(dagSize uint64, entries <-chan unixfs.Li
...
+	//
+	// Range notes:
+	//   - Generating response to a range request may require additional data
+	//     beyond the passed ranges (e.g. a single byte range from the middle of a
+	//     file will still need magic bytes from the very beginning for content
+	//     type sniffing).
+	//   - A range request for a directory currently holds no semantic meaning.
+	//
+	// [HTTP Byte Ranges]: https://httpwg.org/specs/rfc9110.html#rfc.section.14.1.2
+	Get(context.Context, ImmutablePath, ...ByteRange) (ContentPathMetadata, *GetResponse, error)
```
Comment (lidel) [inline gateway/gateway.go:108]: :information_source: removed `GetRange`, we now have `Get` with optional range requests

**Review by lidel — COMMENTED** [review]
@aschmahmann @hacdias I think this is ready for review.

TLDR: adds metrics, fixes range requests, merges `Get` and `GetRange` into one func.

Sharness in Kubo PR is green: https://github.com/ipfs/kubo/pull/9786

**Review by hacdias — APPROVED** [review]
Thanks for this. I tested it locally and it indeed solves the range requests. It LGTM. Nice call to move the metrics stuff to a separate file. It was quite a clutter!
